cpp/src/arrow/compute/exec/tpch_node.cc

```diff
@@ -0,0 +1,3431 @@
+    return Status::OK();
+  }
+
+  Status StartProducing(size_t /*num_threads*/, OutputBatchCallback output_callback,
+                        FinishedCallback finished_callback,
+                        ScheduleCallback /*schedule_task_callback*/) override {
+    std::shared_ptr<Buffer> N_NATIONKEY_buffer =
+        Buffer::Wrap(kNationKey, sizeof(kNationKey));
+    ArrayData N_NATIONKEY_arraydata(int32(), kRowCount,
+                                    {nullptr, std::move(N_NATIONKEY_buffer)});
+
+    ARROW_ASSIGN_OR_RAISE(std::unique_ptr<Buffer> N_NAME_buffer,
+                          AllocateBuffer(kRowCount * kNameByteWidth));
+    char* N_NAME = reinterpret_cast<char*>(N_NAME_buffer->mutable_data());
+    for (size_t i = 0; i < kRowCount; i++)
+      std::strncpy(N_NAME + kNameByteWidth * i, kCountryNames[i], kNameByteWidth);
+    ArrayData N_NAME_arraydata(fixed_size_binary(kNameByteWidth), kRowCount,
+                               {nullptr, std::move(N_NAME_buffer)});
+
+    std::shared_ptr<Buffer> N_REGIONKEY_buffer =
+        Buffer::Wrap(kRegionKey, sizeof(kRegionKey));
+    ArrayData N_REGIONKEY_arraydata(int32(), kRowCount,
+                                    {nullptr, std::move(N_REGIONKEY_buffer)});
+
+    ARROW_ASSIGN_OR_RAISE(Datum N_COMMENT_datum,
+                          g_text.GenerateComments(kRowCount, 31, 114, rng_));
+
+    std::vector<Datum> fields = {
+        std::move(N_NATIONKEY_arraydata), std::move(N_NAME_arraydata),
+        std::move(N_REGIONKEY_arraydata), std::move(N_COMMENT_datum)};
+
+    std::vector<Datum> result;
+    for (const int& col : column_indices_) result.push_back(fields[col]);
+    ARROW_ASSIGN_OR_RAISE(ExecBatch batch, ExecBatch::Make(std::move(result)));
+    output_callback(std::move(batch));
+    finished_callback(static_cast<int64_t>(1));
+    return Status::OK();
+  }
+
+  std::shared_ptr<Schema> schema() const override { return schema_; }
+
+ private:
+  random::pcg32_fast rng_;
+
+  static constexpr size_t kRowCount = 25;
+  static constexpr int32_t kNameByteWidth = 25;
+  const int32_t kNationKey[kRowCount] = {0,  1,  2,  3,  4,  5,  6,  7,  8,
+                                         9,  10, 11, 12, 13, 14, 15, 16, 17,
+                                         18, 19, 20, 21, 22, 23, 24};
+  const char* kCountryNames[kRowCount] = {
+      "ALGERIA",      "ARGENTINA",  "BRAZIL",  "CANADA",         "EGYPT",
+      "ETHIOPIA",     "FRANCE",     "GERMANY", "INDIA",          "INDONESIA",
+      "IRAN",         "IRAQ",       "JAPAN",   "JORDAN",         "KENYA",
+      "MOROCCO",      "MOZAMBIQUE", "PERU",    "CHINA",          "ROMANIA",
+      "SAUDI ARABIA", "VIETNAM",    "RUSSIA",  "UNITED KINGDOM", "UNITED STATES"};
+  const int32_t kRegionKey[kRowCount] = {0, 1, 1, 1, 4, 0, 3, 3, 2, 2, 4, 4, 2,
+                                         4, 0, 0, 0, 1, 2, 3, 4, 2, 3, 3, 1};
+
+  struct NATION {
+    enum {
```

Comment:
You can instead write:
```suggestion
  enum class NATION {
```

Reply:
Also perhaps call this NationColumn or NationField?

Reply:
no, i want NATION::NATIONKEY et. al to be implicitly convertible to int, so that I can use it as a name of indices. with enum class you have to add `static_cast` everywhere.